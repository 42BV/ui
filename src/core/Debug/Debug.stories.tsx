import React from 'react';
import { storiesOf } from '@storybook/react';

import { Debug } from './Debug';

storiesOf('core/Debug', module)
  .addParameters({ component: Debug })
  .add('default open', () => {
    const birthday = new Date();
    birthday.setFullYear(birthday.getFullYear() - 42);

    const user = {
      firstName: 'Hank',
      lastName: 'Hankleton',
      age: 42,
      birthday
    };

    return <Debug value={user} />;
  })
  .add('default closed', () => {
    const birthday = new Date();
    birthday.setFullYear(birthday.getFullYear() - 42);

    const user = {
      firstName: 'Hank',
      lastName: 'Hankleton',
      age: 42,
      birthday
    };

    return <Debug value={user} defaultOpen={false} />;
  });
