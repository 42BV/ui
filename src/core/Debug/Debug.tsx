import React from 'react';

import { CardOpenClose } from '../CardOpenClose/CardOpenClose';
import { UIBasePropsWithCSSPropertiesAndChildren, WithValue } from '../types';

type Props<TVal> = {
  /**
   * Optionally whether the debug should start open or closed.
   *
   * Defaults to `true`.
   */
  defaultOpen?: boolean;
} & Partial<UIBasePropsWithCSSPropertiesAndChildren<() => React.ReactNode>> &
  WithValue<TVal>;

/**
 * A component which helps you debug things by rendering them as
 * JSON on the screen.
 *
 * Useful for when debugging forms:
 *
 * @example
 *
 * ```tsx
 * <Form
 *   onSubmit={save}
 *   initialValues={initialValues}
 *   render={({ handleSubmit, submitting, values }) => (
 *     <form onSubmit={handleSubmit}>
 *       <Debug value={values} />
 *     </form>
 *   )}
 * />
 * ```
 */
export function Debug<TVal>({
  value,
  defaultOpen = true,
  ...props
}: Props<TVal>) {
  return (
    <CardOpenClose
      header="DEBUG"
      isOpen={defaultOpen}
      {...props}
      toggle={() => {
        return;
      }}
    >
      {() => <pre>{JSON.stringify(value, null, 2)}</pre>}
    </CardOpenClose>
  );
}
