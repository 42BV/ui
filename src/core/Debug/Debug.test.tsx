import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import { Debug } from './Debug';

describe('Component: Debug', () => {
  function setup({ defaultOpen }: { defaultOpen?: boolean }) {
    const debug = shallow(
      <Debug value={{ property: 'test' }} defaultOpen={defaultOpen} />
    );
    return { debug };
  }

  test('ui', () => {
    const { debug } = setup({});

    expect(toJson(debug)).toMatchSnapshot();
    expect(toJson(debug.props().children())).toMatchSnapshot();
  });

  it('should start closed when open is false', () => {
    const { debug } = setup({ defaultOpen: false });
    expect(debug.props().isOpen).toBe(false);
  });

  it('should start open when open is true', () => {
    const { debug } = setup({ defaultOpen: true });
    expect(debug.props().isOpen).toBe(true);
  });

  it('should start open when open is undefined', () => {
    const { debug } = setup({});
    expect(debug.props().isOpen).toBe(true);
  });

  describe('events', () => {
    it('should toggle isOpen', () => {
      const setIsOpenSpy = jest.fn();
      jest.spyOn(React, 'useState').mockReturnValue([true, setIsOpenSpy]);

      const { debug } = setup({});

      debug.props().toggle();

      expect(setIsOpenSpy).toHaveBeenCalledTimes(1);
      expect(setIsOpenSpy).toHaveBeenCalledWith(false);
    });
  });
});
