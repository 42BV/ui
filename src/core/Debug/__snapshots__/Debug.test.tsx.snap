// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Debug ui 1`] = `
<CardOpenClose
  header="DEBUG"
  isOpen={true}
  toggle={[Function]}
>
  <Component />
</CardOpenClose>
`;

exports[`Component: Debug ui 2`] = `undefined`;
